src/check_result.jl

```diff
@@ -1,3 +1,83 @@
+# For once you have the sensitivity by two methods (e.g  both finite-differencing and  AD)
+# the code here checks it is correct.
+# Goal is to only call `@isapprox` on things that render well
+# Note that this must work well both on Differnetial types and Primal types
+
+"""
+    _check_equal(actual, expected; kwargs...)
+
+`@test`'s  that `actual ≈ expected`, but breaks up data such that human readable results
+are shown on failures.
+All keyword arguments are passed to `isapprox`.
+"""
+function _check_equal(
+    actual::Union{AbstractArray{<:Number}, Number},
+    expected::Union{AbstractArray{<:Number}, Number};
+    kwargs...
+)
+    @test isapprox(actual, expected; kwargs...)
+end
+
+for (T1, T2) in ((AbstractThunk, Any), (AbstractThunk, AbstractThunk), (Any, AbstractThunk))
+    @eval function _check_equal(actual::$T1, expected::$T2; kwargs...)
+        _check_equal(unthunk(actual), unthunk(expected); kwargs...)
+    end
+end
+
+
+function _check_equal(actual::AbstractArray, expected::AbstractArray; kwargs...)
+    if actual == expected  # if equal then we don't need to be smarter
+        @test true
+    else
+        @test eachindex(actual) == eachindex(expected)
+        @testset "$(typeof(actual))[$ii]" for ii in eachindex(actual)
+            _check_equal(actual[ii], expected[ii]; kwargs...)
+        end
+    end
+end
+
+function _check_equal(actual::Composite{P}, expected::Composite{P}; kwargs...) where P
+    if actual == expected  # if equal then we don't need to be smarter
+        @test true
+    else
+        all_keys = union(keys(actual), keys(expected))
+        @testset "$P.$ii" for ii in all_keys
+            _check_equal(getproperty(actual, ii), getproperty(expected, ii); kwargs...)
+        end
+    end
+end
+
+function _check_equal(
+    ::Composite{ActualPrimal}, expected::Composite{ExpectedPrimal}
+    ) where {ActualPrimal, ExpectedPrimal}
+    # this will certainly fail as we have another dispatch for that, but this will give as
+    # good error message
+    @test ActualPrimal === ExpectedPrimal
+end
+
+# This catches comparisons of Composites and Tuples/NamedTuple
+# and gives a error messaage complaining about that
+_check_equal(::C, expected::T) where {C<:Composite, T} = @test C === T
+_check_equal(::T, expected::C) where {C<:Composite, T} = @test C === T
+
+
+_check_equal(::AbstractZero, x; kwargs...) = _check_equal(zero(x), x; kwargs...)
+_check_equal(x, ::AbstractZero; kwargs...) = _check_equal(x, zero(x); kwargs...)
+_check_equal(x::AbstractZero, y::AbstractZero; kwargs...) = @test x === y
```

Comment:
I am a bit dubious about defining these for `DoesNotExist`
what do you think?

Reply:
Hmmm good point. I think I agree that it feels a bit weird to define it. What would the practical utility of doing so be?

Reply:
I have change to just Zero